pkg/kube/secretcontroller/secretcontroller.go

```diff
@@ -271,68 +366,66 @@ var BuildClientsFromConfig = func(kubeConfig []byte) (kube.Client, error) {
+	stop := make(chan struct{})
+	return &Cluster{
+		secretName: secretName,
+		Client:     clients,
+		// access outside this package should only be reading
+		Stop: stop,
+		// for use inside the package, to close on cleanup
+		stop:          stop,
+		InitialSync:   atomic.NewBool(false),
+		SyncTimeout:   &c.remoteSyncTimeout,
 		kubeConfigSha: sha256.Sum256(kubeConfig),
 	}, nil
 }
 
 func (c *Controller) addMemberCluster(secretName string, s *corev1.Secret) {
 	for clusterID, kubeConfig := range s.Data {
-		// clusterID must be unique even across multiple secrets
-		if prev, ok := c.cs.remoteClusters[clusterID]; !ok {
-			log.Infof("Adding cluster_id=%v from secret=%v", clusterID, secretName)
-
-			remoteCluster, err := createRemoteCluster(kubeConfig, secretName)
-			if err != nil {
-				log.Errorf("Failed to add remote cluster from secret=%v for cluster_id=%v: %v",
-					secretName, clusterID, err)
-				continue
-			}
-
-			c.cs.remoteClusters[clusterID] = remoteCluster
-			if err := c.addCallback(remoteCluster.clients, clusterID); err != nil {
-				log.Errorf("Error creating cluster_id=%s from secret %v: %v",
-					clusterID, secretName, err)
-			}
-		} else {
+		action, callback := "Adding", c.addCallback
+		if prev, ok := c.cs.Get(clusterID); ok {
+			action, callback = "Updating", c.updateCallback
+			// clusterID must be unique even across multiple secrets
 			if prev.secretName != secretName {
 				log.Errorf("ClusterID reused in two different secrets: %v and %v. ClusterID "+
 					"must be unique across all secrets", prev.secretName, secretName)
 				continue
 			}
-
 			kubeConfigSha := sha256.Sum256(kubeConfig)
 			if bytes.Equal(kubeConfigSha[:], prev.kubeConfigSha[:]) {
-				log.Infof("Updating cluster_id=%v from secret=%v: (kubeconfig are identical)", clusterID, secretName)
-			} else {
-				log.Infof("Updating cluster %v from secret %v", clusterID, secretName)
-
-				remoteCluster, err := createRemoteCluster(kubeConfig, secretName)
-				if err != nil {
-					log.Errorf("Error updating cluster_id=%v from secret=%v: %v",
-						clusterID, secretName, err)
-					continue
-				}
-				c.cs.remoteClusters[clusterID] = remoteCluster
-				if err := c.updateCallback(remoteCluster.clients, clusterID); err != nil {
-					log.Errorf("Error updating cluster_id from secret=%v: %s %v",
-						clusterID, secretName, err)
-				}
+				log.Infof("%s cluster_id=%v from secret=%v: (kubeconfig are identical)", clusterID, secretName)
+				continue
 			}
 		}
+		log.Infof("%s cluster %v from secret %v", action, clusterID, secretName)
+
+		remoteCluster, err := c.createRemoteCluster(kubeConfig, secretName)
+		if err != nil {
+			log.Errorf("%s cluster_id=%v from secret=%v: %v", action, clusterID, secretName, err)
+			continue
+		}
+		c.cs.Store(clusterID, remoteCluster)
+		if err := callback(clusterID, remoteCluster); err != nil {
+			log.Errorf("%s cluster_id from secret=%v: %s %v", action, clusterID, secretName, err)
+			continue
```

Comment:
IIUC, it has no retry logic.

Reply:
That's how it works currently, the PR is already large so I didn't want to make additional behavior changes.

There are a few possible things to do on failures here:
- when we fail during `add/update` should we run the `delete` callbacks before continuing?
- what if only some of the callbacks failed, we'd end up retrying them all? callbacks would need to be idempotent

Reply:
make sense, a follow up